### PR TITLE
proxy: fix exclude CIDRs

### LINKED ISSUE #76267: ipvs excludeCIDRs seems not work
**What happened**:
When I config kube-proxy with excludeCIDRs, kube-proxy still try to delete my ipvs rules.

**What you expected to happen**:
kube-proxy not delete excludeCIDRs virturl server.

**How to reproduce it (as minimally and precisely as possible)**:
```
[root@iZrj9fs8mk0inonl9l6j26Z ~]# cat kubeadm-config.yaml 
apiVersion: kubeadm.k8s.io/v1beta1
kind: ClusterConfiguration
kubernetesVersion: v1.14.0
controlPlaneEndpoint: "apiserver.cluster.local:6443" # apiserver DNS name
apiServer:
        certSANs:
        - 127.0.0.1
        - apiserver.cluster.local
        - 172.20.241.205
        - 172.20.241.206
        - 172.20.241.207
        - 172.20.241.208
        - 10.103.97.1          # virturl ip
---
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
mode: "ipvs"
ipvs:
        excludeCIDRs: ["10.103.97.1/32"]
```
kubeadm init --config=kubeadm-config.yaml --experimental-upload-certs

Then add a ipvs on node:
```
ipvsadm -Ln
TCP  10.103.97.1:6443 rr
  -> 172.20.241.207:6443          Masq    1      2          0
```
Then join the node, you can see kube-proxy log:
```
graceful_termination.go:160] Trying to delete rs: 10.103.97.1:6443/TCP/172.20.241.207:6443
```
If don't have an active connection, the ipvs will removed.

**Environment**:
- Kubernetes version (use `kubectl version`):
v1.14.0
- OS (e.g: `cat /etc/os-release`):
centos7.4
- Kernel (e.g. `uname -a`):
3.10
- Install tools:
kubeadm

### PR DESCRIPTION
**Pull request by namreg**
**What type of PR is this?**

 /kind bug

**What this PR does / why we need it**:

kube-proxy: Take into account exclude CIDRs while deleting legacy real servers.

**Which issue(s) this PR fixes**:

Fixes #76267

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by kragniz — COMMENTED** [review]
/ok-to-test

**Comment by m1093782566** [comment]
cc @lbernail @Lion-Wei

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1672,14 +1672,17 @@ func (proxier *Proxier) cleanLegacyService(activeServices map[string]bool, curre
...
-			okayToDelete := true
 			rsList, _ := proxier.ipvs.GetRealServers(svc)
 
 			// If we still have real servers graceful termination is not done
 			if len(rsList) > 0 {
-				okayToDelete = false
+				continue
+			}
+
+			if proxier.isIPInExcludeCIDRs(svc.Address) {
```
Comment (lbernail) [inline pkg/proxy/ipvs/proxier.go]: Could we move it at the top of the for loop (and perform this check before doing anything else, to make it clear we completely ignore these CIDRs?)

**Comment by lbernail** [comment]
This PR makes complete sense
Thank you! (and thanks for adding at test case for this!)

**Comment by lbernail** [comment]
I think it will be worth cherry-picking to 1.14, 1.13 and 1.12 (I don't know if continue to support cherry-picking to 1.11)

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (namreg) [reply pkg/proxy/ipvs/proxier.go]: Yeah, of course. It makes sense

**Comment by namreg** [comment]
@lbernail I've created PRs with cherry-pick on 1.12, 1.13 and 1.14. Let me know if we need to create the one on 1.11

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1692,35 +1694,36 @@ func (proxier *Proxier) cleanLegacyService(activeServices map[string]bool, curre
...
 			}
 		}
 	}
 }
 
+func (proxier *Proxier) isIPInExcludeCIDRs(ip net.IP) bool {
+	// make sure it does not fall within an excluded CIDR range.
+	for _, excludedCIDR := range proxier.excludeCIDRs {
+		// Any validation of this CIDR already should have occurred.
+		_, n, _ := net.ParseCIDR(excludedCIDR)
```
Comment (tedyu) [inline pkg/proxy/ipvs/proxier.go]: Shouldn't we check the error return from this call ?

Comment (tedyu) [reply pkg/proxy/ipvs/proxier.go]: Created #76542

Comment (namreg) [reply pkg/proxy/ipvs/proxier.go]: I didn’t investigate this deep, I merely move this code to function. According to the comment above, error should have already be checked at this place. I believe that’s true :) 
/cc @m1093782566 @lbernail

Comment (tedyu) [reply pkg/proxy/ipvs/proxier.go]: I think adding check of error return is good coding practice.

Comment (lbernail) [reply pkg/proxy/ipvs/proxier.go]: Yes it's validated here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/config/validation/validation.go#L261

@m1093782566 will know better if we encourage systematic error checks in kube-proxy, even when we know they are redundant (@tedyu I completely agree on the general statement)

I think the best would probably be to store parsed CIDRs at initialization (but this is out of the scope of this PR)

Comment (tedyu) [reply pkg/proxy/ipvs/proxier.go]: Logged issue #76544

Comment (tedyu) [reply pkg/proxy/ipvs/proxier.go]: Created PR #76548

**Pull request merged**